src/Microsoft.OData.Core/UriParser/Parsers/ODataPathParser.cs

```diff
@@ -450,80 +454,222 @@ private bool TryCreateCountSegment(string segmentText)
-            IEdmPathExpression bindingPath;
-            var targetNavigationSource = this.parsedSegments[parsedSegments.Count - 1].TargetEdmNavigationSource.FindNavigationTarget(navPropSegment.NavigationProperty, BindingPathHelper.MatchBindingPath, this.parsedSegments, out bindingPath);
+        /// <summary>
+        /// Tries to handle the segment as $ref. If it is $ref, then the rest of the path will be parsed/validated in this call.
+        /// </summary>
+        /// <param name="identifier">The identifier that was parsed from this raw segment.</param>
+        /// <param name="parenthesisExpression">The query portion was parsed from this raw segment.
+        /// This value can be null if there is no query portion.</param>
+        /// <returns>Whether the text was $ref.</returns>
+        private bool TryCreateEntityReferenceSegment(string identifier, string parenthesisExpression)
+        {
+            if (!this.IdentifierIs(UriQueryConstants.RefSegment, identifier))
+            {
+                return false;
+            }
 
-            // If we can't compute the target navigation source, then pretend the navigation property does not exist
-            if (targetNavigationSource == null)
+            if (parenthesisExpression != null)
             {
-                throw ExceptionUtil.CreateResourceNotFoundError(navPropSegment.NavigationProperty.Name);
+                throw ExceptionUtil.CreateSyntaxError();
             }
 
-            // Create new NavigationPropertyLinkSegment
-            var navPropLinkSegment = new NavigationPropertyLinkSegment(navPropSegment.NavigationProperty, targetNavigationSource);
+            // The algorithm below looks for the last NavigationPropertySegment before the $ref segment. Whether a NavPropSeg exists
+            // becomes two different code paths:
+            // 1) Backwards compatibility behavior (<= ODL 7.4.x): If the NavPropSeg exists, then it is expected to either be before
+            // KeySegments or before the $ref (i.e. NavPropSeg/KeySeg/KeySeg/.../$ref or NavPropSeg/$ref). Then the NavPropSeg is replaced
+            // with NavigationPropertyLinkSegment. In a previous implementation, if either a NavPropSeg didn't exist before KeySegments or
+            // $ref itself, then this function would throw. This was not correct as $ref can apply to entity sets and similar segments
+            // (e.g. TypeSegment and FilterSegment), and therefore 2) below is implemented to support those scenarios.
+            // 2) If the NavPropSeg does not exist, then this algorithm appends a ReferenceSegment to the existing list of parsed segments.
 
-            // Add this segment to parsedSegments
-            this.parsedSegments.Add(navPropLinkSegment);
+            // Determine the index of the NavigationPropertySegment in either of the following formats:
+            // NavPropSeg/KeySeg/KeySeg/.../$ref or NavPropSeg/$ref

```

Comment:
Can we remove the code to get the last navigation property segment, since above we added code to keep track of the lastNavigationSource?

Reply:
Resolved offline. Will keep as-is due to maintaining backward compat and actual need for last nav prop segment.